selfie.c

```diff
@@ -7765,6 +7899,421 @@ char* smt_binary(char* opt, char* op1, char* op2) {
   return string;
 }
 
+char* smt_ternary(char* opt, char* op1, char* op2, char* op3) {
+  char* string;
+
+  string = string_alloc(1 + string_length(opt) + 1 + string_length(op1) + 1 + string_length(op2) + 1 + string_length(op3) + 1);
+
+  sprintf4(string, "(%s %s %s %s)", opt, op1, op2, op3);
+
+  return string;
+}
+
+uint64_t find_merge_location(uint64_t beq_imm) {
+  uint64_t original_pc;
+  uint64_t original_imm;
+  uint64_t merge_location;
+
+  // assert: the current instruction is a symbolic beq instruction
+  original_pc = pc;
+  original_imm = imm;
+
+  // examine last instruction before target location of the beq instruction
+  pc = pc + (beq_imm - INSTRUCTIONSIZE);
+
+  // we need to know which instruction it is
+  fetch();
+  decode();
+  
+  if (is != JAL)
+    /* no jal instruction -> end of if without else branch
+       merge is directly at target location of the beq instruction possible
+
+    note: this is a dependency on the selfie compiler */
+    merge_location = original_pc + beq_imm;
+  else {
+    if (signed_less_than(imm, 0) == 0) { 
+      /* jal with positive imm -> end of if with else branch
+         we have to skip the else branch in order to merge afterwards
+
+         note: this is a dependency on the selfie compiler
+         the selfie compiler emits a jal instruction with a positive immediate value if it sees an else branch */
+      merge_location = pc + imm;
+
+      pc = original_pc + INSTRUCTIONSIZE;
+    } else
+      /* jal with negative imm -> end of loop body
+         merge is only outside of the loop body possible
+
+         note: this is a dependency on the selfie compiler
+         the selfie compiler emits a jal instruction with a negative immediate value at
+         the end of the loop body in order to jump back to the loop condition */
+      merge_location = pc + INSTRUCTIONSIZE;
+  }
+
+  // we need to check if we are inside of a recursion before we reach the merge location
+  while (pc != merge_location) {
+    fetch();
+    decode();
+
+    if (is == JAL)
+      // if we are inside of a (arbitrarily deep nested) recursion,
+      // we merge only after the entire recursion has been finished (i.e. the program
+      // has reached a program location which is not part of any recursion)
+      if (currently_in_this_procedure(pc + imm)) {
+        if (in_recursion == 0)
+          recursive_merge_location = get_merge_location_from_corresponding_prologue_start(pc + imm);
+
+        merge_location = recursive_merge_location;
+        in_recursion = 1;
+      }
+
+    pc = pc + INSTRUCTIONSIZE;
+  }
+
+  // restore the original program state
+  pc = original_pc;
+  imm = original_imm;
```

Comment:
You don't need to remember `imm`. Decode will restore that, no?

Reply:
Is this true?

Reply:
That is true, I was just going to fix that in the next update.